### DISCUSSION TITLE
Eclipse CDT 2023-03 (4.27.0) randomly crashes on Rocky Linux 9.2

### DISCUSSION THREAD

**Discussion by elfrances** [eclipse-cdt/cdt#469] [Q&A]
I'm experiencing random crashes while running the Eclipse CDT app on my Rocky Linux 9.2 development PC.  By "crash" I mean the app terminates unexpectedly for no apparent reason, as everything else on the PC is working just fine. 

Upon restarting the app there are no signs of the crash.  Are there crash reports stored anywhere that may shed some light?  

I'm using the following java version:

$ java --version
openjdk 11.0.19 2023-04-18 LTS
OpenJDK Runtime Environment (Red_Hat-11.0.19.0.7-2) (build 11.0.19+7-LTS)
OpenJDK 64-Bit Server VM (Red_Hat-11.0.19.0.7-2) (build 11.0.19+7-LTS, mixed mode, sharing)

**Comment by jonahgraham** [comment]
Thanks for the bug report. That is very frustrating and I'll try to help.

Eclipse these days comes bundled with Java in most cases, and CDT has required Java 17 since late 2022. So the Java version on your path is not actually the one being used. 

Here are some things to try, start Eclipse from the terminal (in case you aren't already) so that any panic messages from Java end up visible on stdout in the terminal. Java writes to `hs_err_pid.log`, where `pid` is the PID of the process by default on fatal errors. Start eclipse with `-consoleLog` which shows the contents of the Eclipse log on the console too (although the log is stored in `workspace/.metadata/log`). 

If you suspect that the bundled version of Java with Eclipse is causing the error, you can start Eclipse with your own java install, for example `eclipse -vm $(which java)`. If you find a difference in behaviour with the stock Java VM vs the bundled one we would be very interested to hear that.

My first guess is that there is an fatal error in the native parts of Eclipse and I hope some of the above will indicate where. A fatal error in Java will normally have a native + java stack trace so we can track that down.

**Comment by elfrances** [comment]
Thanks for your prompt response!

This is how Eclipse CDT is being launched by the desktop manager:

[mmourier@rocky ~]$ ps ax | fgrep eclipse
  26575 pts/2    Sl+    0:00 ./eclipse
  26599 pts/2    Sl+    9:02 /home/mmourier/.p2/pool/plugins/org.eclipse.justj.openjdk.hotspot.jre.full.linux.x86_64_17.0.7.v20230425-1502/jre/bin/java -Declipse.p2.max.threads=10 -Doomph.update.url=https://download.eclipse.org/oomph/updates/milestone/latest -Doomph.redirection.index.redirection=index:/->http://git.eclipse.org/c/oomph/org.eclipse.oomph.git/plain/setups/ --add-opens=java.base/java.lang=ALL-UNNAMED --add-opens=java.base/java.io=ALL-UNNAMED --add-opens=java.base/sun.nio.ch=ALL-UNNAMED --add-opens=java.base/java.net=ALL-UNNAMED --add-opens=java.base/sun.security.ssl=ALL-UNNAMED -Dosgi.requiredJavaVersion=17 -Dosgi.instance.area.default=@user.home/eclipse-workspace -Dosgi.dataAreaRequiresExplicitInit=true -Dorg.eclipse.swt.graphics.Resource.reportNonDisposed=true -Dsun.java.command=Eclipse -Xms256m -Xmx2048m -XX:+UseG1GC -XX:+UseStringDeduplication --add-modules=ALL-SYSTEM -jar /home/mmourier/eclipse/cpp-2023-03/eclipse//plugins/org.eclipse.equinox.launcher_1.6.400.v20210924-0641.jar -os linux -ws gtk -arch x86_64 -showsplash /home/mmourier/.p2/pool/plugins/org.eclipse.epp.package.common_4.28.0.20230608-1200/splash.bmp -launcher /home/mmourier/eclipse/cpp-2023-03/eclipse/eclipse -name Eclipse --launcher.library /home/mmourier/.p2/pool/plugins/org.eclipse.equinox.launcher.gtk.linux.x86_64_1.2.700.v20221108-1024/eclipse_11801.so -startup /home/mmourier/eclipse/cpp-2023-03/eclipse//plugins/org.eclipse.equinox.launcher_1.6.400.v20210924-0641.jar --launcher.appendVmargs -exitdata 9 -product org.eclipse.epp.package.cpp.product -vm /home/mmourier/.p2/pool/plugins/org.eclipse.justj.openjdk.hotspot.jre.full.linux.x86_64_17.0.7.v20230425-1502/jre/bin/java -vmargs -Declipse.p2.max.threads=10 -Doomph.update.url=https://download.eclipse.org/oomph/updates/milestone/latest -Doomph.redirection.index.redirection=index:/->http://git.eclipse.org/c/oomph/org.eclipse.oomph.git/plain/setups/ --add-opens=java.base/java.lang=ALL-UNNAMED --add-opens=java.base/java.io=ALL-UNNAMED --add-opens=java.base/sun.nio.ch=ALL-UNNAMED --add-opens=java.base/java.net=ALL-UNNAMED --add-opens=java.base/sun.security.ssl=ALL-UNNAMED -Dosgi.requiredJavaVersion=17 -Dosgi.instance.area.default=@user.home/eclipse-workspace -Dosgi.dataAreaRequiresExplicitInit=true -Dorg.eclipse.swt.graphics.Resource.reportNonDisposed=true -Dsun.java.command=Eclipse -Xms256m -Xmx2048m -XX:+UseG1GC -XX:+UseStringDeduplication --add-modules=ALL-SYSTEM -jar /home/mmourier/eclipse/cpp-2023-03/eclipse//plugins/org.eclipse.equinox.launcher_1.6.400.v20210924-0641.jar

The version of java used to run it is:

[mmourier@rocky ~]$ /home/mmourier/.p2/pool/plugins/org.eclipse.justj.openjdk.hotspot.jre.full.linux.x86_64_17.0.7.v20230425-1502/jre/bin/java --version
openjdk 17.0.7 2023-04-18
OpenJDK Runtime Environment Temurin-17.0.7+7 (build 17.0.7+7)
OpenJDK 64-Bit Server VM Temurin-17.0.7+7 (build 17.0.7+7, mixed mode)

After the last crash (which happened while I did a CTRL-C and CTRL-V to copy-paste a line of code in a .c file) I launched Eclipse manually from a terminal, as you suggested.  The app has not crashed again yet, but I already see some disturbing log messages on the console:

[mmourier@rocky eclipse]$ ./eclipse 
SLF4J: Failed to load class "org.slf4j.impl.StaticLoggerBinder".
SLF4J: Defaulting to no-operation (NOP) logger implementation
SLF4J: See http://www.slf4j.org/codes.html#StaticLoggerBinder for further details.

(Eclipse:26599): Gtk-CRITICAL **: 14:43:02.083: gtk_box_gadget_distribute: assertion 'size >= 0' failed in GtkScrollbar

(Eclipse:26599): Gtk-CRITICAL **: 14:43:02.084: gtk_box_gadget_distribute: assertion 'size >= 0' failed in GtkScrollbar

(Eclipse:26599): Gtk-CRITICAL **: 14:43:02.084: gtk_box_gadget_distribute: assertion 'size >= 0' failed in GtkScrollbar

Another odd thing I noticed is that in the main source file edit window, sometimes the left column that shows the line numbers starts to flicker, or to jitter up and down.  In most cases, using the mouse wheel to scroll the text up or down, makes this flicker/jitter stop ...

**Reply by jonahgraham** [reply]
As ridiculous as it may sound `Gtk-CRITICAL` messages are generally not a problem to worry about. The theme used in Linux can affect how many of those you see.

I have seen some other reports about Eclipse and the line number column, but on its own isn't likely to cause a crash.

The copy paste causing a crash makes sense, but obviously on its own we can't quite reproduce that. I run xubuntu and sometimes have issues with the clipboard, so it is conceivable that one of those could crash the process.

Hopefully if you get a hs err log file that will help identify the crash.

**Comment by elfrances** [comment]
I just noticed I have an hs_err log file on my home directory.  It is 2 days old, so it is not from my latest Eclipse crash I had this morning. But it may still have some clues about the crash.  I have uploaded it to my Google Drive:

https://drive.google.com/file/d/1J5mD37gFG-mVPLBnTfwTMxn2ZGfDwlA3/view?usp=sharing

**Reply by jonahgraham** [reply]
The log file had exactly what was needed to identify the error - sadly it doesn't come with a fix, but maybe a workaround.

The issue you are seeing is https://github.com/eclipse-platform/eclipse.platform.swt/issues/158 - indeed this https://github.com/eclipse-platform/eclipse.platform.swt/issues/158#issuecomment-1492858295 talks about copy-paste being a fully reproducible error.

As for workarounds - it seems the problem is because of Wayland. If you have only recently started seeing these crashes, perhaps you recently updated your distro and it now uses Wayland? There are some suggestions on https://github.com/eclipse-platform/eclipse.platform.swt/issues/158#issuecomment-1235827515

**Reply by jonahgraham** [reply]
as an interesting P.S. if I google "sigsegv cairo_region_num_rectangles+0x4" (the key part of the error extracted from the log) one of the related searches google suggests is "switch from wayland to x11"

**Reply by elfrances** [reply]
Thanks Jonah.  I'm new to Rocky Linux.  In the past I've worked with Centos 8 and Ubuntu 22, where I've never seen this Eclipse CDT crash.  I assume Rocky Linux must use Wayland by default?  I'll figure out how to switch from Wayland to good-old X11, to see if it takes care of the crash...

**Reply by elfrances** [reply]
Turns out that disabling Wayland is very simple.  On my Rocky 9 system you only need to edit /etc/gdm/custom.conf and uncomment the line:  WaylandEnable=false.  After restarting the system, it will use X11 instead of Wayland.

Since I switched to X11 earlier this morning I have not seen the Eclipse CDT crash, but interestingly I have not noticed that weird flickering/jitter in the line number column either.
